cassiopeia/solver/SpectralNeighborJoiningSolver.py

```diff
@@ -0,0 +1,322 @@
+            )
+
+            lambda_matrix_arr[i_idx, j_idx] = lambda_matrix_arr[
+                j_idx, i_idx
+            ] = svd2_val
+
+        np.fill_diagonal(lambda_matrix_arr, np.inf)
+
+        # regenerate lambda matrix
+        lambda_matrix_df = pd.DataFrame(
+            lambda_matrix_arr, index=node_names, columns=node_names
+        )
+
+        return lambda_matrix_df
+
+    def setup_root_finder(self, cassiopeia_tree: CassiopeiaTree) -> None:
+        """Defines the implicit rooting strategy for the
+        SpectralNeighborJoiningSolver.  See 'setup_root_finder' in
+        NeighborJoiningSolver.
+
+        By default, the SpectralNeighborJoining algorithm returns an
+        unrooted tree.  To root this tree, an implicit root of all zeros is
+        added to the character matrix. Then, the dissimilarity map is
+        recalculated using the updated character matrix. If the tree already
+        has a computed dissimilarity map, only the new similarities are
+        calculated.
+
+        Args:
+            cassiopeia_tree: Input CassiopeiaTree to `solve`
+        """
+        character_matrix = cassiopeia_tree.character_matrix.copy()
+        rooted_character_matrix = character_matrix.copy()
+
+        root = [0] * rooted_character_matrix.shape[1]
+        rooted_character_matrix.loc["root"] = root
+        cassiopeia_tree.root_sample_name = "root"
+        cassiopeia_tree.character_matrix = rooted_character_matrix
+
+        if self.dissimilarity_function is None:
+            raise DistanceSolverError(
+                "Please specify a dissimilarity function to add an implicit "
+                "root, or specify an explicit root"
+            )
+
+        dissimilarity_map = cassiopeia_tree.get_dissimilarity_map()
+        if dissimilarity_map is None:
+            cassiopeia_tree.compute_dissimilarity_map(
+                self.dissimilarity_function, self.prior_transformation
+            )
+        else:
+            dissimilarity = {"root": 0}
+            for leaf in character_matrix.index:
+                weights = None
+                if cassiopeia_tree.priors:
+                    weights = solver_utilities.transform_priors(
+                        cassiopeia_tree.priors, self.prior_transformation
+                    )
+                dissimilarity[leaf] = self.dissimilarity_function(
+                    rooted_character_matrix.loc["root"].values,
+                    rooted_character_matrix.loc[leaf].values,
+                    cassiopeia_tree.missing_state_indicator,
+                    weights,
+                )
+            cassiopeia_tree.set_dissimilarity("root", dissimilarity)
+
+        cassiopeia_tree.character_matrix = character_matrix
+
+    def root_tree(
+        self, tree: nx.Graph, root_sample: str, remaining_samples: List[str]
+    ) -> nx.DiGraph:
+        """Roots a tree produced by Neighbor-Joining at the specified root.
+
+        Uses the specified root to root the tree passed in
```

Comment:
slightly hard to understand, could you rephrase slightly?

Reply:
alright so I'm gonna change it to:
Assigns a root to 'tree' based on a given 'root_sample'. 

the phrasing used originally is exactly quoted from the same function in NeighborJoiningSolver.py, should I edit the docstring for that one as well?